cranelift-codegen/src/isa/x86/enc_tables.rs

```diff
@@ -1236,3 +1237,64 @@ fn convert_ineg(
         }
     }
 }
+
+/// For SIMD shift left, convert an `ishl` to a `x86_psll`, moving the shift index to an XMM
+/// register. Note that the conversions `ishl + constant shift indexes -> ishl_imm` should be done
+/// in [cranelift_codegen::simple_preopt].
+fn convert_ishl(
+    inst: ir::Inst,
+    func: &mut ir::Function,
+    _cfg: &mut ControlFlowGraph,
+    _isa: &dyn TargetIsa,
+) {
+    let mut pos = FuncCursor::new(func).at_inst(inst);
+    pos.use_srcloc(inst);
+
+    if let ir::InstructionData::Binary {
+        opcode: ir::Opcode::Ishl,
+        args: [arg0, arg1],
+    } = pos.func.dfg[inst]
+    {
+        let value_type = pos.func.dfg.value_type(arg0);
+        if value_type.is_vector() && value_type.lane_type().is_int() {
+            if value_type.lane_type() == I8 {
+                // Since x86 does not have a lowering for 8x16 shifts and some of the SIMD spec
+                // committee wants to include this instruction, this non-optimal lowering is
+                // provided (see discussion at https://github.com/WebAssembly/simd/issues/117).
+                // It will use a 16x8 shift and mask off the necessary bits from a table of
+                // constants. This may cause latency due to memory access if the constant table is
+                // not in cache.
+                let shift_index = pos.ins().bitcast(I64X2, arg1);
+                let bitcast_vector = pos.ins().raw_bitcast(I16X8, arg0);
+                let shifted_vector = pos.ins().x86_psll(bitcast_vector, shift_index);
+                let mask_base = pos.ins().iconst(I64, 42); // FIXME this needs to be the base pointer of the constant table
```

Comment:
@sunfishcode, @bnjbvr: I don't know how to get a pointer to the table of constants but I need something like that in order to load the mask. Is there a different way to do this? I looked at `binemit.rs::const_disp4` and it is not clear to me if/how at this point (when I'm doing legalizations) I can get an offset to the constant table.

Reply:
@sunfishcode, here's my current thinking on this:
 - I considered whether having some sort of symbol (unrelated to globals) in the CL IR like `$func.rodata` would make this work; if that symbol could have a known address at compile time then I could use that and constant offsets to get at runtime-determined constants.
 - But can the address of that symbol be known at compile time? I think it can if it is function-related (not global) and relative to the PC: 1) once we know how long the function will be, [calculate all of the offsets of jump tables, constants, and (now) symbols](https://github.com/bytecodealliance/cranelift/blob/cabbacbe63477fcc7d1f1e36717d84f08f749a98/cranelift-codegen/src/binemit/relaxation.rs#L130-L143), and 2) use functions like `jt_disp4`, `const_disp4`, and (now) `symbol_disp4` in `binemit.rs` to pull out those calculated offsets when we emit the code.
 - Looking at this from distance, we would be doing the same type of thing three times and the code is fragile due to the distance between when the offsets are assigned (`relax_branches`), referenced in code (`binemit.rs`), and actually populated with data (`emit_function`). Perhaps we add an abstraction to `Function` like `RelativeOffsets` that would map any of these things (jump table, constant, symbol) to their offset; it would only be usable after some new counting pass (removing that logic from `relax_branches`), would be queryable when encoding instructions and would know how to lay out the actual data in `emit_function`